In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

from scipy.stats import normaltest
from sklearn import datasets, preprocessing, metrics
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [4]:
df = pd.read_excel('2012.xls', index_col=None)
df.dropna()

DateTime  Middeltemperatur  Maksimumtemperatur  \
0   2012-04-30 00:00:00              10.1                10.4   
1   2012-04-30 01:00:00               9.8                10.0   
2   2012-04-30 02:00:00               9.6                 9.7   
3   2012-04-30 03:00:00               8.3                 9.2   
4   2012-04-30 04:00:00               8.9                 9.3   
..                  ...               ...                 ...   
160 2012-05-06 20:00:00               6.8                 7.4   
161 2012-05-06 21:00:00               6.2                 6.6   
162 2012-05-06 22:00:00               5.4                 5.9   
163 2012-05-06 23:00:00               3.8                 4.4   
164 2012-05-07 00:00:00               3.1                 3.9   

     Minimumtemperatur  
0                  9.7  
1                  9.5  
2                  9.2  
3                  7.5  
4                  8.2  
..                 ...  
160                6.6  
161                5.8  
162                4.5  
163                3.1  
164                2.3  

[165 rows x 4 columns]

In [5]:
df = df.rename(columns={"Maksimumtemperatur": "Maksimumtemperatur2012"})

In [6]:
df1 = pd.read_excel('2013.xls', index_col=None)
df1.dropna()

DateTime  Middeltemperatur  Maksimumtemperatur  \
0   2013-04-29 00:00:00               6.2                 6.5   
1   2013-04-29 01:00:00               6.1                 6.3   
2   2013-04-29 02:00:00               6.3                 6.5   
3   2013-04-29 03:00:00               6.4                 6.6   
4   2013-04-29 04:00:00               6.4                 6.7   
..                  ...               ...                 ...   
162 2013-05-05 20:00:00              11.4                12.1   
163 2013-05-05 21:00:00              10.8                11.6   
164 2013-05-05 22:00:00               9.6                10.5   
165 2013-05-05 23:00:00               8.3                 8.7   
166 2013-05-06 00:00:00               8.3                 8.9   

     Minimumtemperatur  
0                  6.0  
1                  6.0  
2                  6.1  
3                  6.1  
4                  6.2  
..                 ...  
162               10.5  
163                9.4  
164                8.3  
165                7.8  
166                7.3  

[167 rows x 4 columns]

In [7]:
df1 = df1.rename(columns={"Maksimumtemperatur": "Maksimumtemperatur2013"})

In [8]:
df2 = pd.read_excel('2014.xls', index_col=None)
df2.dropna()

DateTime  Middeltemperatur  Maksimumtemperatur  \
0   2014-05-01 00:00:00              10.6                11.1   
1   2014-05-01 01:00:00               9.4                10.4   
2   2014-05-01 02:00:00               7.9                 8.6   
3   2014-05-01 03:00:00               7.2                 7.6   
4   2014-05-01 04:00:00               7.3                 7.5   
..                  ...               ...                 ...   
712 2014-05-30 20:00:00              16.8                17.7   
713 2014-05-30 21:00:00              16.5                17.1   
714 2014-05-30 22:00:00              15.3                16.3   
715 2014-05-30 23:00:00              14.2                14.8   
716 2014-05-31 00:00:00              13.4                14.0   

     Minimumtemperatur  
0                 10.1  
1                  8.5  
2                  7.3  
3                  7.0  
4                  7.0  
..                 ...  
712               15.6  
713               16.0  
714               14.7  
715               13.7  
716               12.8  

[717 rows x 4 columns]

In [9]:
df2 = df2.rename(columns={"Maksimumtemperatur": "Maksimumtemperatur2014"})

In [10]:
dfresult = pd.concat([df, df1, df2], axis=1)

In [11]:
dfresult

DateTime  Middeltemperatur  Maksimumtemperatur2012  \
0   2012-04-30 00:00:00              10.1                    10.4   
1   2012-04-30 01:00:00               9.8                    10.0   
2   2012-04-30 02:00:00               9.6                     9.7   
3   2012-04-30 03:00:00               8.3                     9.2   
4   2012-04-30 04:00:00               8.9                     9.3   
..                  ...               ...                     ...   
712                 NaT               NaN                     NaN   
713                 NaT               NaN                     NaN   
714                 NaT               NaN                     NaN   
715                 NaT               NaN                     NaN   
716                 NaT               NaN                     NaN   

     Minimumtemperatur            DateTime  Middeltemperatur  \
0                  9.7 2013-04-29 00:00:00               6.2   
1                  9.5 2013-04-29 01:00:00               6.1   
2                  9.2 2013-04-29 02:00:00               6.3   
3                  7.5 2013-04-29 03:00:00               6.4   
4                  8.2 2013-04-29 04:00:00               6.4   
..                 ...                 ...               ...   
712                NaN                 NaT               NaN   
713                NaN                 NaT               NaN   
714                NaN                 NaT               NaN   
715                NaN                 NaT               NaN   
716                NaN                 NaT               NaN   

     Maksimumtemperatur2013  Minimumtemperatur            DateTime  \
0                       6.5                6.0 2014-05-01 00:00:00   
1                       6.3                6.0 2014-05-01 01:00:00   
2                       6.5                6.1 2014-05-01 02:00:00   
3                       6.6                6.1 2014-05-01 03:00:00   
4                       6.7                6.2 2014-05-01 04:00:00   
..                      ...                ...                 ...   
712                     NaN                NaN 2014-05-30 20:00:00   
713                     NaN                NaN 2014-05-30 21:00:00   
714                     NaN                NaN 2014-05-30 22:00:00   
715                     NaN                NaN 2014-05-30 23:00:00   
716                     NaN                NaN 2014-05-31 00:00:00   

     Middeltemperatur  Maksimumtemperatur2014  Minimumtemperatur  
0                10.6                    11.1               10.1  
1                 9.4                    10.4                8.5  
2                 7.9                     8.6                7.3  
3                 7.2                     7.6                7.0  
4                 7.3                     7.5                7.0  
..                ...                     ...                ...  
712              16.8                    17.7               15.6  
713              16.5                    17.1               16.0  
714              15.3                    16.3               14.7  
715              14.2                    14.8               13.7  
716              13.4                    14.0               12.8  

[717 rows x 12 columns]

In [12]:
dfresult = dfresult.drop(columns=['Middeltemperatur', 'Minimumtemperatur'])

In [13]:
dfresult

DateTime  Maksimumtemperatur2012            DateTime  \
0   2012-04-30 00:00:00                    10.4 2013-04-29 00:00:00   
1   2012-04-30 01:00:00                    10.0 2013-04-29 01:00:00   
2   2012-04-30 02:00:00                     9.7 2013-04-29 02:00:00   
3   2012-04-30 03:00:00                     9.2 2013-04-29 03:00:00   
4   2012-04-30 04:00:00                     9.3 2013-04-29 04:00:00   
..                  ...                     ...                 ...   
712                 NaT                     NaN                 NaT   
713                 NaT                     NaN                 NaT   
714                 NaT                     NaN                 NaT   
715                 NaT                     NaN                 NaT   
716                 NaT                     NaN                 NaT   

     Maksimumtemperatur2013            DateTime  Maksimumtemperatur2014  
0                       6.5 2014-05-01 00:00:00                    11.1  
1                       6.3 2014-05-01 01:00:00                    10.4  
2                       6.5 2014-05-01 02:00:00                     8.6  
3                       6.6 2014-05-01 03:00:00                     7.6  
4                       6.7 2014-05-01 04:00:00                     7.5  
..                      ...                 ...                     ...  
712                     NaN 2014-05-30 20:00:00                    17.7  
713                     NaN 2014-05-30 21:00:00                    17.1  
714                     NaN 2014-05-30 22:00:00                    16.3  
715                     NaN 2014-05-30 23:00:00                    14.8  
716                     NaN 2014-05-31 00:00:00                    14.0  

[717 rows x 6 columns]

In [14]:
dfresult.drop(dfresult.tail(20).index,inplace=True)

In [15]:
dfresult

DateTime  Maksimumtemperatur2012            DateTime  \
0   2012-04-30 00:00:00                    10.4 2013-04-29 00:00:00   
1   2012-04-30 01:00:00                    10.0 2013-04-29 01:00:00   
2   2012-04-30 02:00:00                     9.7 2013-04-29 02:00:00   
3   2012-04-30 03:00:00                     9.2 2013-04-29 03:00:00   
4   2012-04-30 04:00:00                     9.3 2013-04-29 04:00:00   
..                  ...                     ...                 ...   
692                 NaT                     NaN                 NaT   
693                 NaT                     NaN                 NaT   
694                 NaT                     NaN                 NaT   
695                 NaT                     NaN                 NaT   
696                 NaT                     NaN                 NaT   

     Maksimumtemperatur2013            DateTime  Maksimumtemperatur2014  
0                       6.5 2014-05-01 00:00:00                    11.1  
1                       6.3 2014-05-01 01:00:00                    10.4  
2                       6.5 2014-05-01 02:00:00                     8.6  
3                       6.6 2014-05-01 03:00:00                     7.6  
4                       6.7 2014-05-01 04:00:00                     7.5  
..                      ...                 ...                     ...  
692                     NaN 2014-05-30 00:00:00                    12.7  
693                     NaN 2014-05-30 01:00:00                    12.0  
694                     NaN 2014-05-30 02:00:00                    11.7  
695                     NaN 2014-05-30 03:00:00                    11.4  
696                     NaN 2014-05-30 04:00:00                    11.9  

[697 rows x 6 columns]

In [16]:
dfresult.dropna()

DateTime  Maksimumtemperatur2012            DateTime  \
0   2012-04-30 00:00:00                    10.4 2013-04-29 00:00:00   
1   2012-04-30 01:00:00                    10.0 2013-04-29 01:00:00   
2   2012-04-30 02:00:00                     9.7 2013-04-29 02:00:00   
3   2012-04-30 03:00:00                     9.2 2013-04-29 03:00:00   
4   2012-04-30 04:00:00                     9.3 2013-04-29 04:00:00   
..                  ...                     ...                 ...   
160 2012-05-06 20:00:00                     7.4 2013-05-05 18:00:00   
161 2012-05-06 21:00:00                     6.6 2013-05-05 19:00:00   
162 2012-05-06 22:00:00                     5.9 2013-05-05 20:00:00   
163 2012-05-06 23:00:00                     4.4 2013-05-05 21:00:00   
164 2012-05-07 00:00:00                     3.9 2013-05-05 22:00:00   

     Maksimumtemperatur2013            DateTime  Maksimumtemperatur2014  
0                       6.5 2014-05-01 00:00:00                    11.1  
1                       6.3 2014-05-01 01:00:00                    10.4  
2                       6.5 2014-05-01 02:00:00                     8.6  
3                       6.6 2014-05-01 03:00:00                     7.6  
4                       6.7 2014-05-01 04:00:00                     7.5  
..                      ...                 ...                     ...  
160                    15.9 2014-05-07 17:00:00                    14.9  
161                    14.6 2014-05-07 19:00:00                    13.4  
162                    12.1 2014-05-07 20:00:00                    12.0  
163                    11.6 2014-05-07 21:00:00                    10.3  
164                    10.5 2014-05-07 22:00:00                    10.3  

[165 rows x 6 columns]

In [17]:
dfresult.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 697 entries, 0 to 696
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   DateTime                165 non-null    datetime64[ns]
 1   Maksimumtemperatur2012  165 non-null    float64       
 2   DateTime                167 non-null    datetime64[ns]
 3   Maksimumtemperatur2013  167 non-null    float64       
 4   DateTime                697 non-null    datetime64[ns]
 5   Maksimumtemperatur2014  697 non-null    float64       
dtypes: datetime64[ns](3), float64(3)
memory usage: 32.8 KB


In [18]:
dfresult.isnull().sum()

DateTime                  532
Maksimumtemperatur2012    532
DateTime                  530
Maksimumtemperatur2013    530
DateTime                    0
Maksimumtemperatur2014      0
dtype: int64

In [19]:
dfresult.describe()

Maksimumtemperatur2012  Maksimumtemperatur2013  Maksimumtemperatur2014
count              165.000000              167.000000              697.000000
mean                11.260000                9.212575               13.088953
std                  3.626071                3.290845                3.824003
min                  3.900000                1.700000                3.900000
25%                  8.700000                6.700000               10.400000
50%                 10.600000                9.000000               12.500000
75%                 13.800000               11.550000               15.600000
max                 19.100000               16.600000               24.500000

In [20]:
final = dfresult.dropna()

In [21]:
final

DateTime  Maksimumtemperatur2012            DateTime  \
0   2012-04-30 00:00:00                    10.4 2013-04-29 00:00:00   
1   2012-04-30 01:00:00                    10.0 2013-04-29 01:00:00   
2   2012-04-30 02:00:00                     9.7 2013-04-29 02:00:00   
3   2012-04-30 03:00:00                     9.2 2013-04-29 03:00:00   
4   2012-04-30 04:00:00                     9.3 2013-04-29 04:00:00   
..                  ...                     ...                 ...   
160 2012-05-06 20:00:00                     7.4 2013-05-05 18:00:00   
161 2012-05-06 21:00:00                     6.6 2013-05-05 19:00:00   
162 2012-05-06 22:00:00                     5.9 2013-05-05 20:00:00   
163 2012-05-06 23:00:00                     4.4 2013-05-05 21:00:00   
164 2012-05-07 00:00:00                     3.9 2013-05-05 22:00:00   

     Maksimumtemperatur2013            DateTime  Maksimumtemperatur2014  
0                       6.5 2014-05-01 00:00:00                    11.1  
1                       6.3 2014-05-01 01:00:00                    10.4  
2                       6.5 2014-05-01 02:00:00                     8.6  
3                       6.6 2014-05-01 03:00:00                     7.6  
4                       6.7 2014-05-01 04:00:00                     7.5  
..                      ...                 ...                     ...  
160                    15.9 2014-05-07 17:00:00                    14.9  
161                    14.6 2014-05-07 19:00:00                    13.4  
162                    12.1 2014-05-07 20:00:00                    12.0  
163                    11.6 2014-05-07 21:00:00                    10.3  
164                    10.5 2014-05-07 22:00:00                    10.3  

[165 rows x 6 columns]

In [22]:
train_data = final.sample(frac=0.8, random_state=1)
test_data = final.drop(train_data.index)

# Definere mine input variable og output
X_train = train_data[['Maksimumtemperatur2012', 'Maksimumtemperatur2013']]
y_train = train_data['Maksimumtemperatur2014']
X_test = test_data[['Maksimumtemperatur2012', 'Maksimumtemperatur2013']]
y_test = test_data['Maksimumtemperatur2014']

# Laver en linear regression model
model = LinearRegression()

# Træner en med trænings data
model.fit(X_train, y_train)

# Tester modellen på testing dataen
y_pred = model.predict(X_test)

# Evaluaere performance på modellen ved brug af "mean squared error"
mse = mean_squared_error(y_test, y_pred)
print('Mean squared error:', mse)

# Bruger den trænede model til at predicte maximum temperaturen for 2015
X_new = pd.DataFrame({'Maksimumtemperatur2012': ['19'], 'Maksimumtemperatur2013': ['17']})
y_new = model.predict(X_new)
print('Predicted maximum temperature for 2015:', y_new)

Mean squared error: 2.1902124411051096
Predicted maximum temperature for 2015: [15.28613074]


#### in this example i am tpredicting the temperature in Celsius degrees and my MSE is 2.19, that means my model's predictions on average deviate by about 1.48 degrees Celsius from the true values.